nemo/tron/utils/instantiate_utils.py

```diff
@@ -0,0 +1,358 @@
+    args: Tuple[Any, ...],
+    kwargs: Dict[str, Any],
+    full_key: str,
+) -> Any:
+    """Call target (type) with args and kwargs."""
+    args, kwargs = _extract_pos_args(args, kwargs)
+    if _partial_:
+        try:
+            return functools.partial(_target_, *args, **kwargs)
+        except Exception as e:
+            msg = f"Error in creating partial({_convert_target_to_string(_target_)}, ...) object:" + f"\n{repr(e)}"
+            if full_key:
+                msg += f"\nfull_key: {full_key}"
+            raise InstantiationException(msg) from e
+    else:
+        try:
+            return _target_(*args, **kwargs)
+        except Exception as e:
+            msg = f"Error in call to target '{_convert_target_to_string(_target_)}':\n{repr(e)}"
+            if full_key:
+                msg += f"\nfull_key: {full_key}"
+            raise InstantiationException(msg) from e
+
+
+def _convert_target_to_string(t: Any) -> Any:
+    if callable(t):
+        return f"{t.__module__}.{t.__qualname__}"
+    else:
+        return t
+
+
+def _prepare_input_dict_or_list(d: Union[Dict[Any, Any], List[Any]]) -> Any:
+    res: Any
+    if isinstance(d, dict):
+        res = {}
+        for k, v in d.items():
+            if k == "_target_":
+                v = _convert_target_to_string(d["_target_"])
+            elif isinstance(v, (dict, list)):
+                v = _prepare_input_dict_or_list(v)
+            res[k] = v
+    elif isinstance(d, list):
+        res = []
+        for v in d:
+            if isinstance(v, (list, dict)):
+                v = _prepare_input_dict_or_list(v)
+            res.append(v)
+    else:
+        assert False
+    return res
+
+
+def _resolve_target(
+    target: Union[str, type, Callable[..., Any]],
+    full_key: str,
+    check_callable: bool = True,
+) -> Union[type, Callable[..., Any], object]:
+    """Resolve target string, type or callable into type or callable."""
+    if isinstance(target, str):
+        try:
+            target = _locate(target)
+        except Exception as e:
+            msg = f"Error locating target '{target}'."
+            if full_key:
+                msg += f"\nfull_key: {full_key}"
+            raise InstantiationException(msg) from e
+    if check_callable and not callable(target):
+        msg = f"Expected a callable target, got '{target}' of type '{type(target).__name__}'"
+        if full_key:
+            msg += f"\nfull_key: {full_key}"
+        raise InstantiationException(msg)
+    return target
+
+
+def _extract_pos_args(input_args: Any, kwargs: Any) -> Tuple[Any, Any]:
+    config_args = kwargs.pop(_Keys.ARGS, ())
+    output_args = config_args
+
+    if isinstance(config_args, Sequence):
+        if len(input_args) > 0:
+            output_args = input_args
+    else:
+        raise InstantiationException(
+            f"Unsupported _args_ type: '{type(config_args).__name__}'. value: '{config_args}'"
+        )
+
+    return output_args, kwargs
+
+
+def instantiate(
```

Comment:
nit: helpful to have the main API be at the top of the file for readers so they read this first